network/network-service/server/src/tests.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{NetworkServiceNetworkEvents, NetworkServiceServer, NETWORK_SERVER_VERSION};
+use aptos_config::{
+    config::{NetworkServiceConfig, PeerRole},
+    network_id::{NetworkId, PeerNetworkId},
+};
+use aptos_logger::Level;
+use aptos_types::{network_address::NetworkAddress, PeerId};
+use channel::aptos_channel;
+use futures::channel::oneshot;
+use netcore::transport::ConnectionOrigin;
+use network::{
+    application::{
+        storage::PeerMetadataStorage,
+        types::{PeerError, PeerInfo, PeerState},
+    },
+    peer_manager::PeerManagerNotification,
+    protocols::{
+        network::NewNetworkEvents,
+        rpc::InboundRpcRequest,
+        wire::handshake::v1::{MessagingProtocolVersion, ProtocolId, ProtocolIdSet},
+    },
+    transport::{ConnectionId, ConnectionMetadata},
+};
+use network_service_types::{
+    ConnectedPeersResponse, NetworkServiceError, NetworkServiceMessage, NetworkServiceRequest,
+    NetworkServiceResponse, ServerProtocolVersionResponse,
+};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    str::FromStr,
+    sync::Arc,
+};
+
+#[tokio::test]
+async fn test_get_server_protocol_version() {
+    // Create the network client and server
+    let (mut mock_client, service, _) = MockClient::new();
+    tokio::spawn(service.start());
+
+    // Process a request to fetch the protocol version
+    let request = NetworkServiceRequest::GetServerProtocolVersion;
+    let response = mock_client.send_request(request).await.unwrap();
+
+    // Verify the response is correct
+    let expected_response =
+        NetworkServiceResponse::ServerProtocolVersion(ServerProtocolVersionResponse {
+            version: NETWORK_SERVER_VERSION,
+        });
+    assert_eq!(response, expected_response);
+}
```

Comment:
This level of testing feels like there's too much boiler plate code. Is there anyway we can separate the extra layers of networking code with some other generic framework and start routing that through our test frameworks? Testing out all the other logic or rewriting it for each test seems... awful.

Reply:
Yes, because when someone highlights the wrong code, it really can be confusing. Effectively in each piece of code involving networking we need to re-invent some testing mock code to help funnel messages between endpoints. I would propose that we come up with the right abstraction so that we can stop this craziness. That would either imply we'd have a common 100+ lines of code that can be borrowed abstractly or we don't even need, because the code is abstracted at the right level.

Reply:
Gotcha, I've created a Github issue for this (https://github.com/aptos-labs/aptos-core/issues/1208). While I feel it's worthwhile, providing topology awareness feels higher priority to me and so I want to focus on this first. We can always come back and clean things up later 😄